lpvartypes_array.pas

```diff
@@ -759,6 +763,46 @@ function TLapeType_DynArray.Eval(Op: EOperator; var Dest: TResVar; ALeft, ARight
     if wasConstant then
       Result.Writeable := False;
   end
+  else if (op = op_Plus) and (BaseType = ltDynArray) and CompatibleWith(ARight.VarType) then
+  begin 
+    {!FIXME:
+      when `Result := StaticArray + DynArray` result-type is static array, causing this to fail.
+      On the other hand I am unsure why (left) StaticArray enters here.
+      - slacky
+    }
+    Result := NullResVar;
+    Result.VarType := Self;
+    FCompiler.getDestVar(Dest, Result, op);
+
+    if (Result.VarPos.MemPos = mpStack) then
+    begin
+      tmpVar := FCompiler.getTempVar(Self);
+      Result := _ResVar.New(tmpVar);
+    end;
+
+    wasConstant := not Result.Writeable;
+    if wasConstant then Result.Writeable := True;
+
+    Result := Eval(op_Assign, tmpResVar, Result, ALeft, [], Offset, Pos);
+
+    IndexVar := _ResVar.New(FCompiler.getTempVar(ltInt32));
+    with TLapeTree_InternalMethod_Insert.Create(FCompiler, Pos) do
+    try
+      addParam(TLapeTree_ResVar.Create(ARight, FCompiler, Pos));
+      addParam(TLapeTree_ResVar.Create(Result.IncLock(2), FCompiler, Pos));
```

Comment:
A very good question! ;)
I found that if I didn't incLock(2)  it would fail when I did expr alà `result := arr1 + arr2 + arr3`, without being able to come up with any explanation.